tests/vector_tools/integrate_difference_02_nan.cc

```diff
@@ -0,0 +1,129 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2003 - 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// Test integrate_difference
+
+#include "../tests.h"
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_system.h>
+
+using namespace dealii;
+
+
+template <int dim>
+class Ref : public Function<dim>
+{
+public:
+  Ref()
+    :Function<dim>(dim)
+  {}
+
+  double value (const Point<dim> &p, const unsigned int c) const
+  {
+    if (c==0)
+      return numbers::signaling_nan<double>();
+    if (c==1)
+      return p[0]*p[0]+p[1]*p[1];
+    if (c==2)
+      return numbers::signaling_nan<double>();
+    return 0.0;
+  }
+};
+
+
+
+template <int dim>
+void test(VectorTools::NormType norm, double exp = 2.0)
+{
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(2);
+
+  FESystem<dim> fe(FE_Q<dim>(4),dim);
+  DoFHandler<dim> dofh(tria);
+  dofh.distribute_dofs(fe);
+
+  // create a zero vector. the values in there are not important, just
+  // that the difference between this and later the exact solution
+  // happens to have signaling_nans in it
+  Vector<double> solution (dofh.n_dofs ());
+
+  Vector<double> cellwise_errors (tria.n_active_cells());
+  QIterated<dim> quadrature (QTrapez<1>(), 5);
+
+  const ComponentSelectFunction<dim> component_mask (1, 3);
+  VectorTools::integrate_difference (dofh,
+                                     solution,
+                                     Functions::ZeroFunction<dim>(dim),
```

Comment:
Shouldn't this be your function `Ref`? Otherwise, I don't see where `Ref` is used?

Reply:
Oh dear, what have I done. The test clearly doesn't make any sense any more...

I'll get back to you when I have that sorted out.